### ISSUE TITLE
Non deterministic training of LSTM with recurrent_dropout>0

### ISSUE THREAD

**Issue by ftesser** [keras-team/keras#17983]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras):
- OS Platform and Distribution: Linux Ubuntu 22.04.2 LTS
- TensorFlow installed from:  binary
- TensorFlow version: v2.12.0-rc1-12-g0db597d0d75 2.12.0
- Python version: 3.10.6
- Exact command to reproduce: run the code below

**Describe the problem**.
The training of a LSTM model with *recurrent_droput>0* is not deterministic: infact setting the op_determinism configuration  (https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_op_determinism) different runs of creation and train of the same model produce different model (different training history).
In contrast, deterministic training is obtained with *recurrent_dropout=0* and also also with the *dropout>0*.

The code below has been used to reproduce the bug.
Note that the  `det_session` function is used each time before the creation of the model.
This function contains the suggested setting from https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_op_determinism, but I have tested this code, with the same results, also with the following version in which suggested determinism sets from various issues or sites  have been added: 

```python
def det_session():
    os.environ['PYTHONHASHSEED'] = str(1)
    rn.seed(1)
    np.random.seed(1)
    tf.random.set_seed(1)
    tf.keras.utils.set_random_seed(1)
    tf.config.experimental.enable_op_determinism()
```
**Describe the current behavior**.
The training of a LSTM model with *recurrent_droput>0* is not deterministic.

**Describe the expected behavior**.
The training of a LSTM model with *recurrent_droput>0* should be deterministic.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? no

**Standalone code to reproduce the issue**.
```python
import tensorflow as tf
from tensorflow.keras.losses import mean_squared_error
from tensorflow.keras.models import Sequential
from tensorflow.keras.optimizers import RMSprop


def det_session():
    tf.keras.utils.set_random_seed(1)
    tf.config.experimental.enable_op_determinism()


def create_model(inputs_shape):
    # Define the model
    model = Sequential()
    model.add(tf.keras.layers.LSTM(4, input_shape=(inputs_shape[1], inputs_shape[2]), dropout=0.0, recurrent_dropout=0.1))
    # Compile the model
    model.compile(optimizer=RMSprop(1e-3), loss=mean_squared_error)
    # Give a summary
    model.summary()
    return model


if __name__ == '__main__':
    inputs = tf.random.normal([32, 10, 8])
    outputs = tf.random.normal([32, 4])

    # Create and train the model the first time
    det_session()
    model = create_model(inputs.shape)
    history_0 = model.fit(inputs, outputs, epochs=10)
    # Create and train the model more times and check the loss history
    for _ in range(10):
        # Create and train the model again and check the history loss
        det_session()
        model = create_model(inputs.shape)
        history = model.fit(inputs, outputs, epochs=10)
        assert history.history['loss'] == history_0.history['loss'], 'Losses history does not corresponds'

```
The issue was first open on tensorflow repo (https://github.com/tensorflow/tensorflow/issues/60170).
The gist here (https://colab.research.google.com/gist/tilakrayal/740c36db1b70b0946d60e7984a61f254/untitled1059.ipynb).

**Source code / logs**.

**Comment by tilakrayal** [comment]
@sachinprasadhs,
I was able to reproduce the issue on tensorflow v2.12 and tf-nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/98517f7983dbdb63506a552656def220/untitled1074.ipynb).

**Comment by ftesser** [comment]
@mihirparadkar thanks for the analisis, however it is strange to me that this code with float32 precision produces deterministic results with `recurrent_dropout=0` and `dropout>0`, so it seems that the regular dropout is deterministic but the recurrent droput is not deterministic. Why?

**Comment by divyashreepathihalli** [comment]
@ftesser The recurrent dropout layer can be more susceptible to numerical error accumulation than the regular dropout layer, because the recurrent hidden state is updated during back-propagation. As a result, if the dropout probability is too high, so is the numerical error accumulation. 

The regular dropout layer does not have numerical error accumulation because it randomly drops out units in the input or hidden layers during training, which means that the input data is still fixed and not affected by dropout during forward propagation. Therefore, there is no accumulated error that affects the input data.

During backward propagation, the regular dropout layer also randomly drops out the same units that were dropped out during forward propagation. This means that the same units that were set to zero during forward propagation will also receive zero gradients during backward propagation, ensuring that the network's weights are updated consistently.
